Insert yorick labels at the start of each block and insert compile-time tracer switch.

ptersilie opened the pull request:
This PR adds yorick labels to basic MIR blocks which should allow us match them against traces in the future.

Though this PR is likely not ready to be merged yet, I have opened it to make it easier to discuss what needs to be done next.

ptersilie committed:
Insert yorick labels at the start of each block

vext01 commented:
I think the plan of action is as follows:

 * Rename all instances of `no_trace`, to `no_sw_trace`.
 * Add a compile time switch which allows us to select which kind of tracing to use.
 * Compile-time guard all software tracing code so it only gets compiled in if we choose software tracing.
 * Make adjustments to this PR hooking guarding the hardware-tracing code in a similar fashion.
 * Merge this PR.

Sounds good?

The first item on the list is an easy one ;)

ltratt commented:
A dumb question: is the convention in rustc to use `XXX` or `FIXME`?

vext01 commented:
You have to use `FIXME` or the tidy checker will get upset. It rejects `XXX` (which is unfortunate, as I prefer `XXX`).

ptersilie committed:
Rename no_trace to no_sw_trace.

vext01 committed:
Always encode MIR when a Yorick tracer is being used.

vext01 committed:
Convert existing uses of YK_NO_SIR to YK_TRACER.

vext01 committed:
Only turn on AddYkSWTCalls MIR transform for software tracing.

ptersilie committed:
Toggle generation of yk labels

Only generate yk labels if hardware tracing was enabled via the
environment variable YK_TRACER.

vext01 commented:
This is now ready for review.

On top of Lukas' label changes, we now have a unified `YK_TRACER` environment variable to (compile-time) choose which tracer to use.

In the long run (not now) we probably want this to be a proper (tracked) compiler switch, as this would cause cargo to rebuild stuff if the tracer choice changes.

ltratt commented:
Is it worth rebasing/squashing this before review?

vext01 commented:
Ready for review.

ltratt commented on src/librustc_codegen_llvm/builder.rs:
```diff
@@ -154,12 +156,52 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
     }
 
+    fn add_yk_block_label(&mut self, di_sp: &DIScope, lbl_name: CString) {
+
```
Blank line.

ltratt commented on src/librustc_codegen_llvm/builder.rs:
```diff
@@ -154,12 +156,52 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
+        // ykrt uses this at runtime to map virtual addresses to MIR blocks.
+        let first_instr = self.first_instruction(self.llbb());
+
+        if first_instr.is_some() {
```
```rust
if let Some(first_instr) = self.first_instruction(...) {
}
```

ltratt commented on src/librustc_metadata/encoder.rs:
```diff
@@ -65,6 +66,20 @@ macro_rules! encoder_methods {
+    match env::var("YK_TRACER") {
+        Ok(val) => {
+            match val.as_str() {
+                "hw" | "sw" | "sw-rustc" => true,
```
Why do we have `sw` and `sw-rustc`?

vext01 replied:
It's explained here:
https://github.com/softdevteam/ykrustc/pull/41/files#diff-1aeb4fd1a8c45243dc9f0a52c1dafc36R8

In short, it's a mode which inserts calls to the trace recorder, but which doesn't encode SIR. We use this for compiling ykrustc itself (we never want to trace the compiler binaries, but we do want to trace it's libraries).

ltratt replied:
OK, got it.

vext01 replied:
In hindsight, I wonder if `sw-nosir` would have been a better name.

ltratt replied:
I tend to agree that's a better name. We can just add a commit at the end to that effect: I don't think we need to edit all the existing commits.

ltratt commented on src/librustc_codegen_ssa/mir/block.rs:
```diff
@@ -803,7 +804,16 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         debug!("codegen_block({:?}={:?})", bb, data);
 
-        if bx.cx().has_debug() {
+        let tracer = match env::var("YK_TRACER") {
```
That's the third time (I think) that this block of code appears somewhere. Should we find some way of putting it in a single place such that it can be called? I appreciate that this might not be easy.

vext01 replied:
There's a comment to this effect somewhere.

The *correct* thing to do is to have a compiler flag, e.g. a `-C tracer=X`. This has other advantages too: if we make the flag tracked, cargo will rebuild things if the flag changes.

Currently we just did the dumbest thing possible to get us going.

vext01 replied:
Here's the comment I mentioned:
https://github.com/softdevteam/ykrustc/pull/41/files#diff-24c5c945888bb0d041e769bfb852de6cR1105

ptersilie replied:
I could maybe add a `FIXME` here...?

vext01 replied:
I don't think it's necessary to duplicate the existing `FIXME`.

I'm going to need to know whether we are going to use a flag or an environment for my CI tweaks though.

ltratt replied:
Just to check: we're saying that because it's an environment variable we can't place this in a single function somewhere?

vext01 replied:
> Just to check: we're saying that because it's an environment variable we can't place this in a single function somewhere?

We can. I've started a branch to use a proper flag. It makes the code much nicer. We can cherry pick my commit into this PR once I'm done.

ltratt replied:
Do you want to do that as a separate PR?

vext01 replied:
We can do if you like.

ltratt replied:
I'm happy to do whatever's easiest for you. You choose.

vext01 replied:
Let's do it as a separate PR then.

ltratt replied:
OK.

ltratt commented:
@ptersilie Please squash.

ptersilie committed:
Rename YK_TRACER mode `sw-rustc` to `sw-nosir`.

ptersilie commented:
Squashed and also added a commit renaming `sw-rustc` to `sw-nosir`.

vext01 commented:
@ptersilie grep the whole tree for `sw-rustc`. There's at least one instance still present:
https://github.com/ptersilie/ykrustc/blob/yk_labels/.buildbot.sh#L12

ptersilie commented:
Oh, I wonder why `ag` didn't show me this!

ltratt commented:
Please squash.

ptersilie commented:
Squashed!

ltratt commented:
bors r+

vext01 commented:
@ptersilie ah I see you've met the tidy checker :)

```
tidy check
tidy error: /home/buildbot-worker8/worker8/buildbot-build-script/build/src/librustc_mir/transform/add_yk_swt_calls.rs:170: line longer than 100 chars
tidy error: /home/buildbot-worker8/worker8/buildbot-build-script/build/src/librustc_codegen_ssa/mir/block.rs:818: line longer than 100 chars
tidy error: /home/buildbot-worker8/worker8/buildbot-build-script/build/src/librustc_codegen_llvm/builder.rs:181: line longer than 100 chars
tidy error: /home/buildbot-worker8/worker8/buildbot-build-script/build/src/librustc_codegen_llvm/builder.rs:192: line longer than 100 chars
tidy error: /home/buildbot-worker8/worker8/buildbot-build-script/build/src/rustllvm/RustWrapper.cpp:958: line longer than 100 chars
```

vext01 commented:
(you can invoke this by running `./x.py test` and waiting no longer than about 10 seconds)

ptersilie commented:
Ah, do I need to fix these manually? Or can I use clippy or something?

vext01 commented:
It's manual. The compiler codebase doesn't use clippy or rustfmt.

ptersilie commented:
Try this.

vext01 commented:
LGTM

ltratt commented:
Please squash.

ptersilie commented:
Squashed!

ltratt commented:
bors r+
